catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/MetacardApplication.java

```diff
@@ -0,0 +1,687 @@
+        });
+
+        exception(RuntimeException.class, (ex, req, res) -> {
+            LOGGER.debug("Exception occured.", ex);
+            res.status(404);
+            res.header(CONTENT_TYPE, APPLICATION_JSON);
+            res.body(util.getJson(ImmutableMap.of("message",
+                    "Could not find what you were looking for")));
+        });
+    }
+
+    private void revertMetacard(Metacard versionMetacard, String id)
+            throws SourceUnavailableException, IngestException {
+        LOGGER.trace("Reverting metacard [{}] to version [{}]", id, versionMetacard.getId());
+        Metacard revertMetacard = MetacardVersionImpl.toMetacard(versionMetacard, types);
+        Action action = Action.fromKey((String) versionMetacard.getAttribute(MetacardVersion.ACTION)
+                .getValue());
+
+        // Only deleted; deleted_content will have already made metacard so we just want to update
+        if (action.equals(Action.DELETED)) {
+            catalogFramework.create(new CreateRequestImpl(revertMetacard));
+        } else {
+            tryUpdate(4, () -> {
+                catalogFramework.update(new UpdateRequestImpl(id, revertMetacard));
+                return true;
+            });
+        }
+    }
+
+    private void revertContentandMetacard(Metacard latestContent, Metacard versionMetacard,
+            String id)
+            throws SourceUnavailableException, IngestException, ResourceNotFoundException,
+            IOException, ResourceNotSupportedException, FederationException,
+            UnsupportedQueryException {
+        LOGGER.trace(
+                "Reverting content and metacard for metacard [{}]. \nLatest content: [{}] \nVersion metacard: [{}]",
+                id,
+                latestContent.getId(),
+                versionMetacard.getId());
+        Map<String, Serializable> properties = new HashMap<>();
+        properties.put("no-default-tags", true);
+        ResourceResponse latestResource = catalogFramework.getLocalResource(new ResourceRequestById(
+                latestContent.getId(),
+                properties));
+
+        ContentItemImpl contentItem = new ContentItemImpl(id,
+                new ByteSourceWrapper(() -> latestResource.getResource()
+                        .getInputStream()),
+                latestResource.getResource()
+                        .getMimeTypeValue(),
+                latestResource.getResource()
+                        .getName(),
+                latestResource.getResource()
+                        .getSize(),
+                MetacardVersionImpl.toMetacard(versionMetacard, types));
+
+        // Try to delete the "deleted metacard" marker first.
+        Action action = Action.fromKey((String) versionMetacard.getAttribute(MetacardVersion.ACTION)
+                .getValue());
+        if (DELETE_ACTIONS.contains(action)) {
+            // TODO (RCZ) - Semantics for failing the delete? want to still let the recreate happen?
```

Comment:
Na that was just left over thought processes from design stage. will remove :+1: